admin/invoice.php

```diff
@@ -124,7 +133,39 @@ function couponpopupWindow(url) { /* just a stub for coupon output that might fi
             <th class="dataTableHeadingContent">&nbsp;</th>
             <th class="dataTableHeadingContent"><?php echo TABLE_HEADING_PRODUCTS; ?></th>
             <th class="dataTableHeadingContent"><?php echo TABLE_HEADING_PRODUCTS_MODEL; ?></th>
+<?php
+          // -----
+          // Additional column-headings can be added before the Tax columns.
+          //
+          // A watching observer can provide an associative array in the following format (for the products' listing ONLY):
+          //
+          // $extra_headings = array(
+          //     array(
+          //       'align' => $alignment,    // One of 'center', 'right', or 'left' (optional)
+          //       'text' => $value
+          //     ),
+          // );
+          //
+          // Observer notes:
+          // - Be sure to check that the $p2/$extra_headings value is specifically (bool)false before initializing, since
+          //   multiple observers might be injecting content!
+          // - If heading-columns are added, be sure to add the associated data columns, too, via the
+          //   'NOTIFY_ADMIN_INVOICE_DATA_B4_TAX' notification.
+          //
+          $extra_headings = false;
+          $zco_notifier->notify('NOTIFY_ADMIN_INVOICE_HEADING_B4_TAX', '', $extra_headings);
+          if (is_array($extra_headings)) {
+              foreach ($extra_headings as $heading_info) {
+                  $align = (isset($heading_info['align'])) ? (' text-' . $heading_info['align']) : '';
+?>
+            <th class="dataTableHeadingContent<?php echo $align; ?>"><?php echo $heading_info['text']; ?></th>
+<?php
+              }
+          }
+?>
+<?php if ($show_product_tax) { ?>
```

Comment:
While I'm not necessarily a fan of requiring all variables to be included in an observer, I don't see in this file where `$show_product_tax` is ever offered an opportunity to not be `true` other than changing (hard coding) the "default" setting above. I say this because, in one respect variables such as `$extra_headings` are included in the latest notifier, both that variable and `$show_product_tax` are accessible as global variables...

So what is the intent of adding the test for `$show_product_tax` if not to support its adjustment within a notifier?

Reply:
Agreed. I see no reason why it should not be included. I did the mod against 1.5.7 first. $show_product_tax was not in that version. I can add it. My store in the UK does not charge tax (under the vat threshold and do not sell to EU at less that the collection point) so it could simplify my invoices if it were alterable. The only fields that I wanted to alter myself without hardcoding where $show_product_images and $show_attrib_images as I currently hardcode them on my site.

If this were to be added, then would have to move line 26 '$show_product_tax = true;' to above the notifier

if global variable are to be used then the notifier could be removed or changed to ADMIN_INVOICE_START and nothing passed. '$show_product_tax = true;' would still need moving,

Reply:
I was curious about the possibility that the setting may literally be product based (tax on this product, but not on another). Hadn't looked at how the placement might be modified based on that perspective and its possible that others have more reason to nix the discussion. E.g. if it should be an all or nothing per store, per order, per product, etc..., then placement and consideration should end up being in the right place if observer modifiable at all.

Thanks for explaining at least how it came to be a part of the recommended change.